### PR TITLE
[Lockfile] Store what spec repo a pod comes from

### PR DESCRIPTION
**Pull request by segiddins**
For https://github.com/CocoaPods/CocoaPods/pull/7358

### CODE REVIEW

File: lib/cocoapods-core/lockfile.rb [hunk 1]
```diff
@@ -440,6 +453,13 @@ def generate_dependencies_data(podfile)
         podfile.dependencies.map(&:to_s).sort
       end
 
+      def generate_spec_repos(spec_repos)
```
Comment (dnkoutso) [inline lib/cocoapods-core/lockfile.rb]: nit if you care: add doc for this method

File: lib/cocoapods-core/lockfile.rb [hunk 2]
```diff
@@ -440,6 +453,13 @@ def generate_dependencies_data(podfile)
...
+        spec_repos.map do |source, specs|
+          next unless source
+          [source.url || source.name, specs.map(&:name)]
+        end.compact.to_h
```
Comment (segiddins) [inline lib/cocoapods-core/lockfile.rb]: `to_h` doesn't exist on 2.0.0

Comment (dnkoutso) [reply lib/cocoapods-core/lockfile.rb]: <3

File: lib/cocoapods-core/lockfile.rb (hunk 1 above)
Comment (segiddins) [reply lib/cocoapods-core/lockfile.rb]: should this filter out sources with no specs? probably

File: lib/cocoapods-core/yaml_helper.rb [hunk 3]
```diff
@@ -143,6 +145,8 @@ def process_array(array)
       # @return [String] the YAML representation of the given object.
       #
       def process_hash(hash, hash_keys_hint = nil, line_separator = "\n")
+        return '{}' if hash.empty?
```
Comment (segiddins) [inline lib/cocoapods-core/yaml_helper.rb]: this breaks "empty" lockfiles, so we might need to add a test for empty lockfiles explicitly?

Comment (segiddins) [reply lib/cocoapods-core/yaml_helper.rb]: no it doesn't, since we always add the CocoaPods version. The change in the integration spec is expected (and correct), since we're manually removing the cocoapods version (to avoid noisy diffs)

**Comment by segiddins** [comment]
Feedback addressed

**Comment by dnkoutso** [comment]
💯
